server/src/main/java/org/elasticsearch/cluster/metadata/MetadataUpdateSettingsService.java

```diff
@@ -56,223 +56,261 @@ public class MetadataUpdateSettingsService {
-                || (isWildcard && normalizedSettings.hasValue(key) == false)
-                : "unknown setting: " + key + " isWildcard: " + isWildcard + " hasValue: " + normalizedSettings.hasValue(key);
-            settingsForClosedIndices.copy(key, normalizedSettings);
-            if (isWildcard || setting.isDynamic()) {
-                settingsForOpenIndices.copy(key, normalizedSettings);
-            } else {
-                skippedSettings.add(key);
-            }
+    private final class UpdateSettingsTask implements ClusterStateTaskListener {
+        private final UpdateSettingsClusterStateUpdateRequest request;
+        private final ActionListener<AcknowledgedResponse> listener;
+
+        private UpdateSettingsTask(UpdateSettingsClusterStateUpdateRequest request, ActionListener<AcknowledgedResponse> listener) {
+            this.request = request;
+            this.listener = listener;
         }
-        final Settings closedSettings = settingsForClosedIndices.build();
-        final Settings openSettings = settingsForOpenIndices.build();
-        final boolean preserveExisting = request.isPreserveExisting();
-
-        // TODO: move this to custom class instead of AckedClusterStateUpdateTask
-        AckedClusterStateUpdateTask clusterTask = new AckedClusterStateUpdateTask(
-            Priority.URGENT,
-            request,
-            wrapPreservingContext(listener, threadPool.getThreadContext())
-        ) {
-            @Override
-            public ClusterState execute(ClusterState currentState) {
-                RoutingTable.Builder routingTableBuilder = null;
-                Metadata.Builder metadataBuilder = Metadata.builder(currentState.metadata());
-
-                // allow to change any settings to a closed index, and only allow dynamic settings to be changed
-                // on an open index
-                Set<Index> openIndices = new HashSet<>();
-                Set<Index> closedIndices = new HashSet<>();
-                final String[] actualIndices = new String[request.indices().length];
-                for (int i = 0; i < request.indices().length; i++) {
-                    Index index = request.indices()[i];
-                    actualIndices[i] = index.getName();
-                    final IndexMetadata metadata = currentState.metadata().getIndexSafe(index);
-                    if (metadata.getState() == IndexMetadata.State.OPEN) {
-                        openIndices.add(index);
-                    } else {
-                        closedIndices.add(index);
-                    }
+
+        @Override
+        public void onFailure(Exception e) {
+            listener.onFailure(e);
+        }
+
+        @Override
+        public void clusterStateProcessed(ClusterState oldState, ClusterState newState) {
+            assert false : "should not be called";
+        }
+
+        private ClusterStateAckListener getAckListener() {
```

Comment:
Typically ack-sensitive tasks implement `ClusterStateAckListener` themselves rather than creating a new object. Not that it really matters, just a question of harmony with other implementations.